resources/type-system/normalization.md

```diff
@@ -80,7 +80,7 @@ equations apply gives us something like the following:
 - **NORM**(`X extends T`) =
   - let `S` be **NORM**(`T`)
   - if `S` is `Never` then `Never`
-  - else `X extends S`
+  - else `X extends T`
```

Comment:
Why this change? Presumably, `X extends int??` (for example) would be normalized to `X extends int?`, but IIUC this change is undoing that.

Reply:
Note the difference between normalizing a *use* of a type variable and the *declaration* of a type variable.  This case covers type variable uses. It's not clear to me what it means to normalize the bound on a use.  It's fine to do so, if a representation is chosen in which each use has a copy of the bound, but the bound does not contribute to syntactic equality.   The bound does contribute to normalization, since `X extends Never` normalizes to `Never`.  Does that make sense?

Reply:
Sort of. I'm reading the various expressions in this document as (abstract) type terms, and in my mental model, these aren't intrinsically equipped with a notion of being declared vs. used. (In a Dart program, type terms certainly exist in declaration-contexts and use-contexts, but I'm interpreting normalization as a context-free operation.)

Maybe I missed a definition somewhere, but I'm not really sure how to read `X extends T` as a standalone type term. (That is, I recognize it as a *declaration* in the context of a type parameter list, but I'm not sure how to interpret an `extends` clause without that context the way I can interpret `FutureOr<X>`, for example.) In this case, I took it as shorthand for the following: I have a type variable `X` and when I query my type environment's oracle/function/mechanism for looking up bounds, it tells me that **bound**(`X`) = `T`.

Reply:
> In this case, I took it as shorthand for the following: I have a type variable `X` and when I query my type environment's oracle/function/mechanism for looking up bounds, it tells me that **bound**(`X`) = `T`.

Functionally, that is the correct interpretation.  The analyzer's representation doesn't really work that way.  Instead, they attach a bound to every type variable, and there's essentially a coherence property that you enforce.  But it's morally the same thing - it's just a question of whether you keep a map from type variables to bounds, or just enforce that all type variables bound at the same location have the same bound attached to them.  I've just gotten in the habit of using the `X extends T` notation instead of talking about "the bound associated with T in the environment" since that also matches up better with the implementations.